File: compiler/src/bananabread/runtime/Handlers.scala
package bananabread
package runtime

import value.Value
import instruction._
import register._


def handle(code: Code, state: State): Dispatch = code match
  case _: Data   => Cont
  case _: Label  => Cont
  case Halt      => Stop
  case op: FrameInit => Error("unknown operator", op)
  case op: Jz    => handleJz(op, state)
  case op: Jmp   => handleJmp(op, state)
  case op: Push  => handlePush(op, state)
  case op: Call  => handleCall(op, state)
  case Call0     => handleCall0(Call0, state)
  case Ret       => handleRet(Ret, state)
  case Swap      => handleSwap(Swap, state)
  case op: Mov   => handleMov(op, state)
  case op: Stw   => handleStw(op, state)
  case op: Ldw   => handleLdw(op, state)
  case op: Load  => handleLoad(op, state)
  case op: Store => handleStore(op, state)
  case Println   => handlePrintln(Println, state)
  case Concat    => handleConcat(Concat, state)
  case op: Add   => handleAdd(op, state)
  case op: Sub   => handleSub(op, state)
  case op: Frame => handleFrame(op, state)

def handleJz(op: Jz, state: State): Dispatch = state.pop match
  case value.I32(0) => goto(op, op.label, state)
  case _            => Cont

def handleJmp(op: Jmp, state: State): Dispatch =
  goto(op, op.label, state)

def handlePush(op: Push, state: State): Dispatch = op match
  case Push(I32, v: value.I32) =>
    state.push(v)
    Cont
  case Push(Bool, v: value.Bool) =>
    state.push(v)
    Cont
  case Push(Ref, v: value.Id) =>
    state.push(v)
    Cont
  case Push(Scope, value.Id(label)) =>
    val scope = value.Scope(label, state.frames.curr)
    state.push(scope)
    Cont
  case Push(Const, value.Id(label)) =>
    const(op, label, state) { const =>
      state.push(const)
      Cont
    }
  case Push(Ptr, value.Id(label)) =>
    state.heap.mapped(label) match
      case None =>
        Error("bad push: invalid pointer", op)
      case Some(ptr) =>
        state.push(value.I32(ptr.addr))
        Cont
  case _ =>
    Error("bad push", op)

def handleCall(op: Call, state: State): Dispatch = state.frames.curr.get(op.label) match
  case None =>
    state.push(value.I32(state.registers.pc.value + 1))
    state.frames.next
    goto(op, op.label, state)
  case Some(ptr: value.Id) =>
    state.push(value.I32(state.registers.pc.value + 1))
    state.frames.next
    goto(op, ptr.label, state)
  case Some(value.Scope(label, frame)) =>
    state.push(value.I32(state.registers.pc.value + 1))
    state.frames.from(frame)
    goto(op, label, state)
  case Some(bad) =>
    Error(s"bad call: ${op.label} is instance of `${bad.getClass.getSimpleName}`, expected function", op)

def handleCall0(op: Instruction, state: State): Dispatch =
  state.push(value.I32(state.registers.pc.value + 1))
  Jump(state.registers.jm.value)

def handleRet(op: Instruction, state: State): Dispatch = state.pop match
  case value.I32(addr) =>
    state.frames.prev
    Jump(addr)
  case bad =>
    Error(s"bad ret: missing return address: $bad", op)

def handleSwap(op: Instruction, state: State): Dispatch =
  if state.registers.esp.value == 0
  then Cont
  else
    val a = state.pop
    val b = state.pop
    state.push(a)
    state.push(b)
    Cont

def handleMov(op: Mov, state: State): Dispatch = op match
  case Mov(reg, Some(addr), Some(offset), None) =>
    state.heap.mapped(addr.label) match
      case Some(ptr) =>
        val data = state.heap.lookup(ptr.addr + offset.value)
        state.registers.setValue(reg, data)
        Cont
      case None =>
        Error(s"bad mov: missing label ${addr.label}", op)
  case Mov(reg, None, Some(offset), None) =>
    ???
  case Mov(reg, Some(addr), None, Some(offsetReg)) =>
    (state.heap.mapped(addr.label), state.registers.getValue(offsetReg)) match
      case (Some(ptr), offset: value.I32) =>
        val data = state.heap.lookup(ptr.addr + offset.value)
        state.registers.setValue(reg, data)
        Cont
      case (Some(ptr), v) =>
        Error(s"bad mov: ${offsetReg} register did not contain a valid offset", op)
      case (None, _) =>
        Error(s"bad mov: missing label ${addr.label}", op)
  case Mov(reg, None, None, None) => state.pop match
    case addr: value.I32 =>
      state.registers.set(reg, addr)
      Cont
    case value.Id(label) =>
      state.labels.get(label) match
        case Some(i) =>
          state.registers.set(reg, i)
          Cont
        case None =>
          Error(s"bad mov: missing label $label", op)
    case value.Scope(label, frame) =>
      state.labels.get(label) match
        case Some(i) =>
          state.frames.from(frame)
          state.registers.set(reg, value.I32(i))
          Cont
        case None =>
          Error(s"bad mov: missing scope $label", op)
    case _ =>
      Error("bad mov: invalid stack entry", op)

def handleStw(op: Stw, state: State): Dispatch =
  op.reg match
    case Rt =>
      state.push(state.registers.rt)
      Cont
    case Ebp =>
      state.push(state.registers.ebp)
      Cont
    case Esp =>
      state.push(state.registers.esp)
      Cont
    case Rax =>
      state.push(state.registers.rax)
      Cont
    case _ =>
      Error("bad stw: register does not accept user data", op)

def handleLdw(op: Ldw, state: State): Dispatch =
  (op.reg, state.pop) match
    case (Rt | Rax, v) =>
      state.registers.setValue(op.reg, v)
      Cont
    case (Ebp, v : value.I32) =>
      state.registers.ebp(v.value)
      Cont
    case (Esp, v : value.I32) =>
      state.registers.esp(v.value)
      Cont
    case (Ebp | Esp, _) =>
      Error("bad ldw: register does not accept non-int32 data", op)
    case (_, _) =>
      Error("bad ldw: register does not accept user data", op)

def handleLoad(op: Load, state: State): Dispatch = state.frames.curr.get(op.label) match
  case None =>
    pointer(op, op.label, state) { value =>
      state.push(value)
      Cont
    }
  case Some(v) =>
    state.push(v)
    Cont

def handleStore(op: Store, state: State): Dispatch =
  state.frames.curr.put(op.label, state.pop)
  Cont

def handlePrintln(ins: Instruction, state: State): Dispatch =
  println(state.pop)
  Cont

def handleConcat(op: Instruction, state: State): Dispatch =
  binStrOp(state)(_ + _) match
    case Some(v) =>
      state.push(v)
      Cont
    case None =>
      Error("bad concat: missing argument", op)

def handleAdd(op: Add, state: State): Dispatch =
  binI32Op(state)(_ + _) match
    case Some(v) =>
      state.push(v)
      Cont
    case None =>
      Error("bad add: missing argument", op)

def handleSub(op: Sub, state: State): Dispatch =
  binI32Op(state)(_ - _) match
    case Some(v) =>
      state.push(v)
      Cont
    case None =>
      Error("bad sub: missing argument", op)

def handleFrame(op: Frame, state: State): Dispatch =
  Cont

def goto(op: Instruction, label: String, state: State): Dispatch = state.labels.get(label) match
  case None => Error(s"bad jump: missing label: ${label}", op)
  case _ => Goto(label)

def const(op: Instruction, label: String, state: State)(f: Value => Dispatch): Dispatch =
  state.heap.lookup(label) match
    case None => Error(s"missing const: ${label}", op)
    case Some(value) => f(value)

def pointer(op: Instruction, label: String, state: State)(f: Value => Dispatch): Dispatch =
  (state.heap.lookup(label), state.labels.get(label)) match
    case (None, None) => Error(s"invalid pointer: ${label}", op)
    case (Some(value), _) => f(value)
    case (_, Some(_)) => f(value.Id(label))

def binI32Op(state: State)(f: (Int, Int) => Int): Option[value.I32] =
  (state.pop, state.pop) match
    case (value.I32(rhs), value.I32(lhs)) =>
      Some(value.I32(f(lhs, rhs)))
    case _ =>
      None

def binStrOp(state: State)(f: (String, String) => String): Option[value.Str] =
  (state.pop, state.pop) match
    case (value.Str(rhs), value.Str(lhs)) =>
      Some(value.Str(f(lhs, rhs)))
    case _ =>
      None

File: vm/src/handlers.cpp
#include "handlers.hpp"
#include "value.hpp"
#include "instruction.hpp"

#include <stack>
#include <string>
#include <map>

namespace Bananabread {
namespace Handlers {

using std::stack;
using std::string;
using std::map;

Dispatch::Action* handle_push(Instruction::Push* push, State* state) {
  switch (push->get_type()) {
    case Instruction::Value::Type::Invalid:
      return new Dispatch::Error("invalid type for push instruction");
    case Instruction::Value::Type::Str:
      return new Dispatch::Error("unimplemented: push str");
    case Instruction::Value::Type::I32:
      try {
        auto inner = std::stoi(push->get_value());
        auto value = new Value::I32(inner);
        state->stack->push(value);
        return new Dispatch::Cont();
      } catch (std::invalid_argument) {
        return new Dispatch::Error("bad push: invalid i32 value");
      }
    case Instruction::Value::Type::Ref:
      return new Dispatch::Error("unimplemented: push ref");
    case Instruction::Value::Type::Const:
      // TODO Convert constants to runtime values before evaluating expression
      // and handle invalid/missing constants.
      auto value = Value::from_instruction(state->constants[push->get_value()]);
      state->stack->push(value);
      return new Dispatch::Cont();
  }
}

// TODO handle Some(Id), Some(Scope), and Some(Invalid).
Dispatch::Action* handle_call(Instruction::Call* call, State* state) {
  auto ref = state->frames->get_curr()->get(call->get_label());

  if (!ref) {
    state->stack->push(new Value::I32{state->reg->pc() + 1});
    state->frames->move_to_next();
    return new Dispatch::Goto(call->get_label());
  }

  return new Dispatch::Error("bad call");
}

Dispatch::Action* handle_swap(Instruction::Swap* swap, State* state) {
  auto a = state->stack->top();
  state->stack->pop();
  auto b = state->stack->top();
  state->stack->pop();

  state->stack->push(a);
  state->stack->push(b);

  return new Dispatch::Cont();
}

Dispatch::Action* handle_store(Instruction::Store* store, State* state) {
  state->frames->get_curr()->put(store->get_label(), state->stack->top());
  state->stack->pop();
  return new Dispatch::Cont();
}

Dispatch::Action* handle_load(Instruction::Load* load, State* state) {
  auto ref = state->frames->get_curr()->get(load->get_label());

  if (ref) {
    state->stack->push(ref);
    return new Dispatch::Cont();
  }

  auto inst = state->constants[load->get_label()];
  if (!inst) {
    return new Dispatch::Error("missing const: " + load->get_label());
  }

  auto value = Value::from_instruction(inst);
  state->stack->push(value);
  return new Dispatch::Cont();
}

Dispatch::Action* handle_println(Instruction::Println* println, State* state) {
  auto value = state->stack->top();
  if (!value) {
    return new Dispatch::Error("missing stack value");
  }

  std::cout << value->to_string() << std::endl;
  state->stack->pop();
  return new Dispatch::Cont();
}

Dispatch::Action* handle_ret(Instruction::Ret* ret, State* state) {
  auto value = state->stack->top();
  if (!value) {
    return new Dispatch::Error("bad ret: missing return address");
  }

  if (auto addr = dynamic_cast<Value::I32*>(value)) {
    state->stack->pop();
    state->frames->move_to_prev();
    return new Dispatch::Jump(addr->get_int());
  }

  return new Dispatch::Error("bad ret: invalid return address");
}

Dispatch::Action* handle_add(Instruction::Add* add, State* state) {
  auto rhs_value = state->stack->top();
  state->stack->pop();
  auto lhs_value = state->stack->top();
  state->stack->pop();

  if (auto rhs = dynamic_cast<Value::I32*>(rhs_value)) {
    if (auto lhs = dynamic_cast<Value::I32*>(lhs_value)) {
      auto result = lhs->get_int() + rhs->get_int();
      state->stack->push(new Value::I32(result));
      return new Dispatch::Cont();
    } else {
      return new Dispatch::Error("bad add: left hand side is not an int");
    }
  } else {
    return new Dispatch::Error("bad add: right hand side is not an int");
  }
}

Dispatch::Action* handle_sub(Instruction::Sub* sub, State* state) {
  auto rhs_value = state->stack->top();
  state->stack->pop();
  auto lhs_value = state->stack->top();
  state->stack->pop();

  if (auto rhs = dynamic_cast<Value::I32*>(rhs_value)) {
    if (auto lhs = dynamic_cast<Value::I32*>(lhs_value)) {
      auto result = lhs->get_int() - rhs->get_int();
      state->stack->push(new Value::I32(result));
      return new Dispatch::Cont();
    } else {
      return new Dispatch::Error("bad sub: left hand side is not an int");
    }
  } else {
    return new Dispatch::Error("bad sub: right hand side is not an int");
  }
}

Dispatch::Action* handle_jz(Instruction::Jz* jz, State* state) {
  auto value = state->stack->top();
  if (!value) {
    return new Dispatch::Error("bad jz: empty stack");
  }

  auto cond = dynamic_cast<Value::I32*>(value);
  if (!cond) {
    return new Dispatch::Error("bad jz: invalid head value");
  }

  if (cond->get_int() == 0) {
    return new Dispatch::Goto(jz->get_label());
  }

  return new Dispatch::Cont();
}

Dispatch::Action* handle_jmp(Instruction::Jmp* jmp, State* state) {
  return new Dispatch::Goto(jmp->get_label());
}

Dispatch::Action* handle(Instruction::Code* code, State* state) {
  if (dynamic_cast<Instruction::Label*>(code)) {
    return new Dispatch::Cont();
  } else if (dynamic_cast<Instruction::Value*>(code)) {
    return new Dispatch::Cont();
  } else if (dynamic_cast<Instruction::Halt*>(code)) {
    return new Dispatch::Stop();
  } else if (auto push = dynamic_cast<Instruction::Push*>(code)) {
    return handle_push(push, state);
  } else if (auto call = dynamic_cast<Instruction::Call*>(code)) {
    return handle_call(call, state);
  } else if (dynamic_cast<Instruction::Frame*>(code)) {
    return new Dispatch::Cont();
  } else if (auto swap = dynamic_cast<Instruction::Swap*>(code)) {
    return handle_swap(swap, state);
  } else if (auto store = dynamic_cast<Instruction::Store*>(code)) {
    return handle_store(store, state);
  } else if (auto load = dynamic_cast<Instruction::Load*>(code)) {
    return handle_load(load, state);
  } else if (auto println = dynamic_cast<Instruction::Println*>(code)) {
    return handle_println(println, state);
  } else if (auto ret = dynamic_cast<Instruction::Ret*>(code)) {
    return handle_ret(ret, state);
  } else if (auto add = dynamic_cast<Instruction::Add*>(code)) {
    return handle_add(add, state);
  } else if (auto sub = dynamic_cast<Instruction::Sub*>(code)) {
    return handle_sub(sub, state);
  } else if (auto jz = dynamic_cast<Instruction::Jz*>(code)) {
    return handle_jz(jz, state);
  } else if (auto jmp = dynamic_cast<Instruction::Jmp*>(code)) {
    return handle_jmp(jmp, state);
  }

  return new Dispatch::Error("internal error: unhandled instruction");
}

} // namespace Handlers
} // namespace Bananabread
